### PR TITLE
gh-117657: Replace TSAN suppresions with more specific rules

### PR DESCRIPTION
**Pull request by swtaarrs**
Using `race:` filters out warnings if the function appears anywhere in the stack trace. This can hide a lot of unrelated warnings, especially for a function like `_PyEval_EvalFrameDefault`, which is somewhere on the stack more often than not.

Change all free-threaded suppressions to `race_top:`, which only matches the top frame, and add any new suppressions this exposes. I iterated the test suite under TSAN until it could go 5 times in a row without any new warnings.

- Issue: gh-117657

### CODE REVIEW

**Comment by swtaarrs** [comment]
Tests are passing now but @mpage pointed out that I should probably do some more stress-testing of this before landing, since there are likely more warnings that just show up less frequently.

File: Tools/tsan/suppressions_free_threading.txt [hunk 1]
```diff
@@ -11,34 +11,58 @@ race:set_allocator_unlocked
...
+race_top:start_the_world
+race_top:tstate_set_detached
+race_top:unicode_hash
+race_top:Py_SET_TYPE
+race_top:_PyDict_CheckConsistency
+race_top:_PyImport_AcquireLock
+race_top:_Py_atomic_compare_exchange_int
+race_top:_Py_atomic_load_uintptr_relaxed
+race_top:_Py_dict_lookup_threadsafe
+race_top:__tsan_memset
```
Comment (colesbury) [inline Tools/tsan/suppressions_free_threading.txt]: This one (`__tsan_memset`) seems weird. Can we get a function in CPython?

Comment (swtaarrs) [reply Tools/tsan/suppressions_free_threading.txt]: Yeah. I'll also do that for `recv`.

**Comment by colesbury** [comment]
I'm seeing a race sometimes reported in `test_io.test_daemon_threads_shutdown_stdout_deadlock` between `PyThreadState_Clear` and `tstate_activate`:

https://gist.github.com/colesbury/c0a6edcdf7e909191490cedd3873a062

**Comment by colesbury** [comment]
Another race between `type_modified_unlocked` (typeobject.c:960:20) and `subtype_dealloc` (typeobject.c:2296:36) when running test_httpservers.test_authorization.

**Review by mpage — APPROVED** [review]
LGTM!

**Review by colesbury — APPROVED** [review]

**Comment by colesbury** [comment]
The buildbot failure looks unrelated

**Pull request merged**